### PR TITLE
Workflow algorithms for ILL's TOF reduction

### LINKED ISSUE #17382: Data reduction workflow algorithm for TOF instruments at ILL
The planned data reduction workflow for the time-of-flight instruments at ILL is outlined here:
https://github.com/mantidproject/documents/blob/master/Project-Management/ILL/TOF_and_BS/Mantid_workflow_proposal.md

The algorithm is to be written in Python.

### PR DESCRIPTION
**Pull request by soininen**
This PR adds ILL's time-of-flight reduction package to Mantid framework. The changes include six new Python workflow algorithms, their unit tests, documentation as well as changes to relevant instrument parameter and mask files (IN4, IN5 and IN6). No other functionality should have been altered.

In some parts this is still work-in-progress, so there are a few TODO's in the source code. Most notably, usage examples are missing in the documentation. These will be added once the instrument scientists provide a suitable dataset.

**To test:**

Scientific validation and extensive testing has been done by the instrument scientists at ILL. Thus, checking that the builds pass and code review should be enough.

Closes #17382.

*Release notes were updated accordingly, push [here](https://github.com/mantidproject/mantid/blob/04368bd5aaf5f6ce68d2e1548c68aa21858066ed/docs/source/release/v3.11.0/direct_inelastic.rst).*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/DirectILLReduction-v1.rst
Comment (NickDraper) [inline docs/source/algorithms/DirectILLReduction-v1.rst]: Even if you do not have a test from the scientists, could you not create a simple example here based on what you do in the unit tests?  This would at least show the python syntax in the documentation even if it is a fake usage scenario (several others are like that).  This would apply to all of the other usage tests as well.

Comment (NickDraper) [reply docs/source/algorithms/DirectILLReduction-v1.rst]: Of course if the others are purely internal and done by users through this as a primary workflow algorithm the examples might be better on this page.

File: instrument/IN4_Parameters.xml
Comment (NickDraper) [inline instrument/IN4_Parameters.xml]: These changes will go out to users immediately once merged, not with the next release.  The look safe to me, but you might want to double check.

File: instrument/IN5_Parameters.xml
Comment (NickDraper) [inline instrument/IN5_Parameters.xml]: These changes will go out to users immediately once merged, not with the next release.  The look safe to me, but you might want to double check.

File: instrument/IN6_Parameters.xml
Comment (NickDraper) [inline instrument/IN6_Parameters.xml]: These changes will go out to users immediately once merged, not with the next release.  The look safe to me, but you might want to double check.

**Review by NickDraper — CHANGES_REQUESTED** [review]
Coding is good, I'm worried about the usage examples unless completing them gets moved to another issue.

Assuming functionality and scientific assessment has been verified by users as stated.

File: instrument/IN4_Parameters.xml
Comment (soininen) [reply instrument/IN4_Parameters.xml]: I think the users should be happy as the old efficiency corrections for IN5 and IN6 were incorrect.

For IN4, the formulas need a bit more work, as described in #20459.

**Comment by soininen** [comment]
I've created issue #20453 to keep track of the missing usage example data.

**Review by NickDraper — APPROVED** [review]

**Comment by NickDraper** [comment]
Good to go once the build servers are happy

**Pull request merged**
